pandas/_libs/lib.pyx

```diff
@@ -2741,7 +2742,11 @@ def maybe_convert_objects(ndarray[object] objects,
         seen.object_ = True
 
     elif seen.str_:
-        if using_string_dtype() and is_string_array(objects, skipna=True):
+        if (
+            convert_string
```

Comment:
In what cases would we have `convert_string` be False and `using_string_dtype` be True? Not saying this implementation is wrong, just hard to grok at first glance

Reply:
This combination of `convert_string=False` with `using_string_dtype()` being True is exactly what I am using inside the replace/fillna implementation, to pass down to `maybe_convert_objects` that I don't want to cast object to string _if_ I know that we started with object dtype, and that this dtype should be preserved

Reply:
But I fully agree it is all a bit confusing, and the implementation is not very clear (I mostly got to the current state just by getting all our tests to pass, but it is difficult to assess whether it is complete. However, given that this is about "not raising a useless warning", I think it is not the worst thing if we would have missed a case where we would still raise the warning)

Reply:
Ah OK thanks that is helpful. As a nit, maybe `preserve_object_dtype` would be a better keyword, but not something I think is a blocker. This all needs a good cleanup past once we get through the 2.3 push anyway

Reply:
> As a nit, maybe `preserve_object_dtype` would be a better keyword

It's only specifically not converting object to string here (all other inferred types are still used), so `preserve_object_dtype` would be a bit too generic I think